homeassistant/helpers/trigger.py

```diff
@@ -68,6 +73,107 @@ class TriggerInfo(TypedDict):
     trigger_data: TriggerData
 
 
+@dataclass
+class PluggableActionsEntry:
+    """Holder to keep track of all plugs and actions for a given trigger."""
+
+    plugs: set[PluggableAction] = field(default_factory=set)
+    actions: dict[object, tuple[HassJob, dict[str, Any]]] = field(default_factory=dict)
+
+
+class PluggableAction:
+    """A pluggable action handler."""
+
+    _entry: PluggableActionsEntry | None = None
+
+    def __init__(self, update: CALLBACK_TYPE | None = None) -> None:
+        """Initialize a pluggable action.
+
+        :param update: callback triggered whenever triggers are attached or removed.
+        """
+        self._update = update
+
+    def __bool__(self) -> bool:
+        """Return if we have something attached."""
+        return bool(self._entry and self._entry.actions)
+
+    def async_run_update(self) -> None:
+        """Run update function if one exists."""
+        if self._update:
+            self._update()
+
+    @staticmethod
+    def async_get_registry(hass: HomeAssistant) -> dict[tuple, PluggableActionsEntry]:
+        """Return the pluggable actions registry."""
+        if data := hass.data.get(DATA_PLUGGABLE_ACTIONS):
+            return data  # type: ignore[no-any-return]
+        data = defaultdict(PluggableActionsEntry)
+        hass.data[DATA_PLUGGABLE_ACTIONS] = data
+        return data
+
+    @staticmethod
+    def async_attach_trigger(
+        hass: HomeAssistant,
+        trigger: dict[str, str],
+        action: TriggerActionType,
+        variables: dict[str, Any],
+    ) -> CALLBACK_TYPE:
+        """Find all registered plugs of given trigger config."""
+        reg = PluggableAction.async_get_registry(hass)
+        key = tuple(sorted(trigger.items()))
+        entry = reg[key]
+
+        def _update() -> None:
+            for plug in entry.plugs:
+                plug.async_run_update()
+
+        @callback
+        def _remove() -> None:
+            """Remove this action attachment, and disconnect all plugs."""
+            del entry.actions[_remove]
+            _update()
+            if not entry.actions and not entry.plugs:
+                del reg[key]
+
+        job = HassJob(action)
+        entry.actions[_remove] = (job, variables)
+        _update()
+
+        return _remove
+
+    @callback
+    def async_register(
+        self, hass: HomeAssistant, trigger: dict[str, str]
+    ) -> CALLBACK_TYPE:
+        """Register plug in the global plugs dictionary."""
+
+        reg = PluggableAction.async_get_registry(hass)
+        key = tuple(sorted(trigger.items()))
+        self._entry = reg[key]
+        self._entry.plugs.add(self)
+
+        @callback
+        def _remove() -> None:
+            """Remove plug from registration, and make sure no action holds on to disconnection functions."""
+            assert self._entry
+            self._entry.plugs.remove(self)
```

Comment:
The reason for the loose relationship between plugs and actions isn't clear to me. Why do we allow actions that are registered from a specific plug to remain if the plug is removed?

Reply:
We need to remember the actions that was registered when an integration is reloaded. When integration is reloaded, all plugs are unloaded. But the action need to remain. But the comment here is a bit stale. Will revise it.